src/katsdpsigproc/resource.py

```diff
@@ -58,12 +58,22 @@ def ready(*args) -> None:
 async def async_wait_for_events(events: Iterable[AbstractEvent],
                                 loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
     """Coroutine that waits for a list of device events."""
-    def wait_for_events(events: Iterable[AbstractEvent]) -> None:
+    def wait_for_events(events: List[AbstractEvent]) -> None:
         for event in events:
             event.wait()
+        # Remove references to events before the future is resolved. This
+        # prevents a race condition where a caller might await
+        # async_wait_for_events, then drop its references to the events before
+        # the executor's worker thread has a chance to, causing the event to be
+        # destroyed in the worker thread. That in turn leads to a warning from
+        # PyCUDA (and a resource leak) if the context can't be made current in
+        # the worker thread.
+        events.clear()
+
     if loop is None:
         loop = asyncio.get_event_loop()
     if events:
+        events = list(events)
```

Comment:
Is this to force all the events in the iterable to be "evaluated" (for want of a better word)?

Reply:
It's also just occurred to me that Iterable doesn't guarantee that truthiness corresponds to a non-empty sequence. I'll move the `list` conversion above the `if` test.